src/routes/blog/post/personal-chatbot-gpt-4o/+page.markdoc

```diff
@@ -0,0 +1,275 @@
+ * @param {string} fileName
+ * @returns {string} Contents of static/{fileName}
+ */
+export function getStaticFile(fileName) {
+  return fs.readFileSync(path.join(staticFolder, fileName)).toString();
+}
+
+/**
+ * Throws an error if any of the keys are missing from the object
+ * @param {*} obj
+ * @param {string[]} keys
+ * @throws {Error}
+ */
+export function throwIfMissing(obj, keys) {
+  const missing = [];
+  for (let key of keys) {
+    if (!(key in obj) || !obj[key]) {
+      missing.push(key);
+    }
+  }
+  if (missing.length > 0) {
+    throw new Error(`Missing required fields: ${missing.join(', ')}`);
+  }
+}
+```
+
+### Creating the function logic
+
+Now that our chatbot UI and utility functions are ready, we can develop our final function logic. For that, we must first install the `openai` NPM package. Open your terminal in the project directory and run the following command:
+
+```bash
+npm i openai
+```
+
+After that is done, we must visit the `src/main.js` file and replace the existing code with the following:
+
+```jsx
+import OpenAI from 'openai';
+import { getStaticFile, throwIfMissing } from './utils.js';
+
+export default async ({ req, res }) => {
+  throwIfMissing(process.env, ['OPENAI_API_KEY']);
+
+  if (req.method === 'GET') {
+    return res.send(getStaticFile('index.html'), 200, {
+      'Content-Type': 'text/html; charset=utf-8',
+    });
+  }
+
+  try {
+    throwIfMissing(req.body, ['prompt']);
+  } catch (err) {
+    return res.json({ ok: false, error: err.message }, 400);
+  }
+
+  const openai = new OpenAI();
+
+  try {
+    const response = await openai.chat.completions.create({
+      model: 'gpt-4o',
+      max_tokens: parseInt(process.env.OPENAI_MAX_TOKENS ?? '512'),
+      messages: [{ role: 'user', content: req.body.prompt }],
+    });
+    const completion = response.choices[0].message.content;
+    return res.json({ ok: true, completion }, 200);
+  } catch (err) {
+    return res.json({ ok: false, error: 'Failed to query model.' }, 500);
+  }
+};
+```
+
+This function will return our chatbot UI on any `GET` request to the function and will use the prompt sent in the request body and use the OpenAI GPT-4o API to generate and return a response otherwise.
+
+At this point, our project directory structure should look as follows:
+
+```
+.
+├── src/
+│   ├── main.js
+│   └── utils.js
+├── static/
+│   └── index.html
+├── README.md
+├── package-lock.json
+└── package.json
+
+```
+
+### Testing the function
+
+Once you've completed all the aforementioned steps, you can push the code to our GitHub repository, at which point Appwrite Cloud will automatically deploy the changes to your function.
+
+![Appwrite Function deployments](/images/blog/personal-chatbot-gpt-4o/deployment.png)
+
+You can then go ahead and test your function by opening the function domain in your browser.
+
+Here is what an example of this looks like: [apwr.dev/gpt-4o-demo](https://apwr.dev/gpt-4o-demo)
```

Comment:
How are we going to afford to keep this online 🧐 ?

Reply:
Based on usage seen on past similar projects, I have more than enough credits available to keep this running for a while